src/main/java/org/zalando/baigan/service/S3ConfigurationRepository.java

```diff
@@ -34,52 +33,99 @@ public class S3ConfigurationRepository extends AbstractConfigurationRepository {
      * Provides a {@link ConfigurationRepository} that reads configurations from a JSON file stored in a S3 bucket.
      * It refreshes configurations using the default refresh interval (60 seconds)
      *
+     * @deprecated use {@link S3ConfigurationRepository#S3ConfigurationRepository(S3FileLoader)} instead
+     *
      * @param bucketName The name of the bucket
      * @param key        The object key, usually, the "full path" to the JSON file stored in the bucket
      */
+    @Deprecated
     public S3ConfigurationRepository(@Nonnull final String bucketName, @Nonnull final String key) {
         this(bucketName, key, DEFAULT_REFRESH_INTERVAL);
     }
 
     /**
      * Provides a {@link ConfigurationRepository} that reads configurations from a JSON file stored in a S3 bucket.
      *
+     * @deprecated use {@link S3ConfigurationRepository#S3ConfigurationRepository(S3FileLoader, long)} instead
+     *
      * @param bucketName      The name of the bucket
      * @param key             The object key, usually, the "full path" to the JSON file stored in the bucket
      * @param refreshInterval The interval, in seconds, to refresh the configurations. A value of 0 disables refreshing
      *                        <p>
      * @see #S3ConfigurationRepository(String, String)
      */
+    @Deprecated
     public S3ConfigurationRepository(@Nonnull final String bucketName, @Nonnull final String key, final long refreshInterval) {
-        this(bucketName, key, refreshInterval, new ScheduledThreadPoolExecutor(1));
+        this(new S3FileLoader(bucketName, key), refreshInterval, new ScheduledThreadPoolExecutor(1));
     }
 
     /**
      * Provides a {@link ConfigurationRepository} that reads configurations from a JSON file stored in a S3 bucket.
      *
+     * @deprecated Use {@link S3ConfigurationRepository#S3ConfigurationRepository(S3FileLoader, long, ScheduledThreadPoolExecutor)} instead
+     *
      * @param bucketName      The name of the bucket
      * @param key             The object key, usually, the "full path" to the JSON file stored in the bucket
      * @param refreshInterval The interval, in seconds, to refresh the configurations. A value of 0 disables refreshing
      * @param executor        The executor to refresh the configurations in the specified interval
      *                        <p>
      * @see #S3ConfigurationRepository(String, String)
      */
+    @Deprecated
```

Comment:
I am not so sure about this change in terms of public API of the library. These constructor overloads basically are acting as "builders" for library consumers to create instances of the `S3ConfigurationRepository`. 

As a library consumer I _want_ to have easy to use builders, which allow me to focus on what I want from the library, without having to know its implementation details (e.g. `S3FileLoader`). I _trust_ the library that what it does internally is well-tested, i.e. I'd never want to test the internals of a library in my application code.

As a library maintainer I want as least of my code to be considered "public", i.e. being used explicitly/directly by consumers, so that I retain flexibility for future refactorings/improvements -> cohesion, isolation, coupling etc pp

Is there a way we can address your (valid) concern of enabling higher testability _internally_ to the library, while keeping a solid and easy-to-use interface for consumers? For example I see many libraries to model such "builders", explicitly, and allow instance creation only via them, while internally classes are modularized and with that well-tested.

Reply:
> I'd never want to test the internals of a library in my application code

Yes, I agree. I chose the shortest path to make it work. But I agree that `S3FileLoader` is something the user should not know of. 

However, the user has to be enabled to configure the S3 client (which is easiest done if they are able to supply the client). Note that this is not to test the library, but rather to end2end test my application that uses this library (just as I would not want to test the internals of my HTTP client, but still be able to specify the endpoint it calls).

I will rewrite this in some way to address your concerns and keep end2end testability by users.